### PR TITLE
First draft of Passport Model

### PR DESCRIPTION
**Pull request by ionut-arm**
Add initial version of the design for supporting Passport Model. Includes the extensions, the guidelines for using said extensions for both peers, and adjustments to the introduction to make it less background-check-centric.

cc @yogeshbdeshpande

### CODE REVIEW

File: draft-fossati-tls-attestation.md [hunk 1]
```diff
@@ -154,74 +154,108 @@ Several formats for encoding evidence are available, such as:
...
+such encoding, AR4SI {{?I-D.ietf-rats-ar4si}} is being standardized by the RATS
 working group.
 
-This version of the specification defines how to support the background check model
-in  the TLS handshake, such that the details about the attestation technology are
-agnostic to the TLS handshake itself. Later versions of the specification will 
-support the passport model as well. 
+This specification defines how to support the background check model in the TLS
+handshake, such that the details about the attestation technology are agnostic
+to the TLS handshake itself. Similarly, support for the passport model is
```
Comment (yogeshbdeshpande) [inline draft-fossati-tls-attestation.md:163]: I would rephrase it as follows.

This specification supports both the background check and passport model, during TLS Handshake.

In background check model the details about the attestation technology are agnostice to TLS handshake itself.
Similarly, in the passport model, the details about the attestation results encoding and trust relationships are agnostic to  the TLS handshake.

File: draft-fossati-tls-attestation.md [hunk 2]
```diff
@@ -154,74 +154,108 @@ Several formats for encoding evidence are available, such as:
...
+defined such that details about the attestation results encodings and trust
+relationships are agnostic to the TLS handshake.
+
+To give the peer information that the handshake signing key is properly secured,
+the associated attestation result has to be appraised by the peer. This must be
+the case when either of the two remote attestation topologies is used. Hence,
+attestation evidence about the security state of the signing key is needed,
+which is typically associated with evidence about the overall platform state.
+The platform attestation service ensures that the key attestation service has
+not been tampered with. The platform attestation service issues the Platform
```
Comment (yogeshbdeshpande) [inline draft-fossati-tls-attestation.md:173]: Hence the attestation evidence (to be appraised by an attestation verifier )
must contain the security state of the signing key and also the security of the 
overall platform state.

File: draft-fossati-tls-attestation.md [hunk 3]
```diff
@@ -154,74 +154,108 @@ Several formats for encoding evidence are available, such as:
...
+handshake, such that the details about the attestation technology are agnostic
+to the TLS handshake itself. Similarly, support for the passport model is
+defined such that details about the attestation results encodings and trust
+relationships are agnostic to the TLS handshake.
+
+To give the peer information that the handshake signing key is properly secured,
+the associated attestation result has to be appraised by the peer. This must be
+the case when either of the two remote attestation topologies is used. Hence,
+attestation evidence about the security state of the signing key is needed,
+which is typically associated with evidence about the overall platform state.
```
Comment (yogeshbdeshpande) [inline draft-fossati-tls-attestation.md:171]: Hence the attestation evidence (to be appraised by an attestation verifier )
must contain the security state of the signing key and also the security of the
overall platform state.

File: draft-fossati-tls-attestation.md [hunk 4]
```diff
@@ -154,74 +154,108 @@ Several formats for encoding evidence are available, such as:
...
+CertificateVerify message, when client-based authentication is requested. The
+attestation payload must contain assertions relating to the client's TLS
+Identity Key (TIK), which associate the private key with the attestation
+information. These assertions may come in the form of a Key Attestation Token
+(KAT), or of specific claims in an attestation result document. An example of a
+KAT format utilizing the EAT format can be found in {{I-D.bft-rats-kat}}.
+
+The relying party can obtain and appraise the remote attestation results either
+directly from the Certificate message (in the passport model), or by relaying
+the evidence from the Certificate message to the verifier. Subsequently, the
```
Comment (yogeshbdeshpande) [inline draft-fossati-tls-attestation.md:230]: I would re-phrase it to;

 or by relaying the evidence from the Certificate message to the verifier and receiving the attestation results. Subsequently, the ...

File: draft-fossati-tls-attestation.md [hunk 5]
```diff
@@ -154,74 +154,108 @@ Several formats for encoding evidence are available, such as:
...
+
+When using the passport model, the remote attestation results obtained by the
+attester from its trusted verifiers can be cached and used for any number of
+subsequent TLS handshakes. This allows the two sides to amortize the cost of
+remote attestation by reducing the frequency with which evidence must be
+produced and appraised. However, this does create a risk of stale attestation
+results being used for authentication. The relying party must therefore be
+empowered to reject an authentication attempt with results that do not meet its
+staleness policy. This should trigger a new evidence appraisal on the attester
+side.
```
Comment (yogeshbdeshpande) [inline draft-fossati-tls-attestation.md:241]: Suggested edit:
When using the passport model, the remote attestation results obtained by the
attester from its trusted verifiers can be cached and used for any number of
subsequent TLS handshakes. This allows the two sides to amortize the cost of
remote attestation by reducing the frequency with which remote attestation needs to be
performed with the verifier.  However, this does create a risk of stale attestation
results being used for authentication. The relying party must therefore be
empowered to reject an authentication attempt with results that do not meet its
acceptance policy. This should trigger a new attestation protocol between attester 
and the verifier.

Comment (yogeshbdeshpande) [reply draft-fossati-tls-attestation.md:241]: Also a generic question, we should specify 
(a) how existing protocols built in the TLS system allow us to reject the Attestation results

(b) If a suitable cause needs to be added in the Reject Message, what the new value should be?

File: draft-fossati-tls-attestation.md [hunk 6]
```diff
@@ -303,16 +337,74 @@ Auth | {CertificateVerify}
 ~~~~
 {: #figure-background-check-model2 title="TLS Client Providing Evidence to TLS Server."}
 
-# Evidence Extensions (Background Check Model)
+## TLS Client authenticating using results
```
Comment (yaronf) [inline draft-fossati-tls-attestation.md:340]:
```suggestion
## TLS Client authenticating using attestation results 
```

File: draft-fossati-tls-attestation.md [hunk 7]
```diff
@@ -303,16 +337,74 @@ Auth | {CertificateVerify}
...
 {: #figure-background-check-model2 title="TLS Client Providing Evidence to TLS Server."}
 
-# Evidence Extensions (Background Check Model)
+## TLS Client authenticating using results 
+
+In this use case the TLS client, as the attester, provides attestation results
+to the TLS server. The TLS client is the attester and the the TLS server acts as
+a relying party. Prior to delivering its Certificate message, the client must
+contact the verifier (not shown in the diagram) to receive the attestation
+results that it will use as credentials.
```
Comment (yaronf) [inline draft-fossati-tls-attestation.md:346]: Here again the 2nd sentence duplicates the first one.

File: draft-fossati-tls-attestation.md [hunk 8]
```diff
@@ -303,16 +337,74 @@ Auth | {CertificateVerify}
...
+                               <--------  [Application Data*]
+     ^ {Certificate}
+Auth | {CertificateVerify}
+     v {Finished}              -------->
+       [Application Data]      <------->  [Application Data]
+~~~~
+{: #figure-passport-model1 title="TLS Client Providing Results to TLS Server."}
+
+
+## TLS Server authenticating using results
```
Comment (yaronf) [inline draft-fossati-tls-attestation.md:374]:
```suggestion
## TLS Server authenticating using attestation results
```

File: draft-fossati-tls-attestation.md [hunk 9]
```diff
@@ -469,11 +561,40 @@ to hashing, the binder must be encoded as described in {{binding-mech}}.
 The hash algorithm negotiatied within the handshake must be used wherever
 hashing is required for the binder.
 
+# Results Extensions (Passport Model)
```
Comment (yaronf) [inline draft-fossati-tls-attestation.md:564]:
```suggestion
# Attestation Results Extensions (Passport Model)
```

File: draft-fossati-tls-attestation.md [hunk 10]
```diff
@@ -852,14 +1053,20 @@ values carried in these extensions are taken from TBD.
 
 IANA is requested to allocate a value in the "TLS Alerts"
```
Comment (yaronf) [inline draft-fossati-tls-attestation.md:1054]: IANA: I think we're also adding two more extensions, right?

**Review by yaronf — CHANGES_REQUESTED** [review]

File: diagrams/passport-tls-handshake.txt [hunk 11]
```diff
@@ -1,29 +1,29 @@
-.-------.      .------------.          .-------.       .----------.
-| TLS A |      | Verifier X |          | TLS B |       | Attester |
-'---+---'      '-----+------'          '---+---'       '----+-----'
+.--------.     .------------.          .--------.      .----------.
+| Client |     | Verifier X |          | Server |      | Attester |
```
[the remark this answers is not in the document]
Comment (ionut-arm) [reply diagrams/passport-tls-handshake.txt:2]: I think it's only meant to highlight the fact that multiple Verifiers might be relevant or present to the client and/or server, and that this is just one of them. Perhaps I can "rename" it to "Verifier A" instead.

File: draft-fossati-tls-attestation.md (hunk 7 above)
Comment (ionut-arm) [reply draft-fossati-tls-attestation.md:346]: I'll likely rebase this PR on top of the early exporters one, so will take care of this afterwards.

File: diagrams/passport-tls-handshake.txt (hunk 11 above)
[the remark this answers is not in the document]
Comment (yaronf) [reply diagrams/passport-tls-handshake.txt:2]: But then again, maybe we should be consistent with "Attester" and remove the "X"?

[the remark this answers is not in the document]
Comment (ionut-arm) [reply diagrams/passport-tls-handshake.txt:2]: Oh, I forgot to fix the "Attester" - in this case the "Attester" is just the Server's "Attestation Service", basically the same entity in the context of TLS, so there's a 1-to-1 mapping between "Server" and "Attester". However, there is a 1-to-many relationship between Server (or Client) and Verifier(s).

Something else I should probably do is to make line 12 read something like `trusted_verifiers([Verifier X, Verifier Y, ...])` to emphasize this.

File: draft-fossati-tls-attestation.md (hunk 3 above)
Comment (muhammad-usama-sardar) [reply draft-fossati-tls-attestation.md:171]: attestation verifier is not used otherwise in the draft, so probably skip "attestation". 

Also, slightly shorter and concise version: 

Hence, the attestation evidence (to be appraised by the verifier)
must contain the security state of the signing key as well as the overall platform.

File: draft-fossati-tls-attestation.md [hunk 12]
```diff
@@ -154,74 +154,108 @@ Several formats for encoding evidence are available, such as:
...
+{: vspace="0"}
+
+TLS Identity Key (TIK): : A cryptographic key used by one of the peers to
+authenticate itself during the TLS handshake.
+
+The reader is assumed to be familiar with the vocabulary and concepts defined in
+{{-rats-arch}}, and those in {{-rats-kat}}.
+
+"Remote attestation evidence" is more succintly referred to as "evidence", and
+"remote attestation results" is more succintly referred to as "results"
```
Comment (muhammad-usama-sardar) [inline draft-fossati-tls-attestation.md:194]: "results" is a very generic word and can be confusing at times. I would suggest "attestation results".

File: draft-fossati-tls-attestation.md [hunk 13]
```diff
@@ -154,74 +154,108 @@ Several formats for encoding evidence are available, such as:
...
+throughout this document. "Remote attestation credentials", or "credentials", is
```
Comment (muhammad-usama-sardar) [inline draft-fossati-tls-attestation.md:195]: Similar to above, I would suggest "attestation credentials" rather than "credentials" alone.

File: draft-fossati-tls-attestation.md [hunk 14]
```diff
@@ -154,74 +154,108 @@ Several formats for encoding evidence are available, such as:
...
-to verify the CertificateVerify message.
-
-# Use of Evidence with the Background Check Model
-
-The background check model is described in Section 5.2 of 
-{{I-D.ietf-rats-architecture}} and allows the following modes
-of operation when used with TLS, namely:
+As typical with new features in TLS, the client indicates support for the new
+extension in the ClientHello message. The newly introduced extensions allow
+remote attestation credentials and nonces to be exchanged. The nonces are used
```
Comment (muhammad-usama-sardar) [inline draft-fossati-tls-attestation.md:212]: Why does nonce need to be mentioned separately? Is it not a part of the "remote attestation credentials"? For example, KAT includes nonce, right?

File: draft-fossati-tls-attestation.md [hunk 15]
```diff
@@ -154,74 +154,108 @@ Several formats for encoding evidence are available, such as:
...
+modes of operation are allowed when used with TLS, namely:
 
 - TLS client is the attester, 
 - TLS server is the attester, and
 - TLS client and server mutually attest towards each other. 
 
-We will show the message exchanges of the three cases in
-sub-sections below.
+We will show the message exchanges of the first two cases in sub-sections below.
+Mutual authentication via attestation combines these two (non-interfering)
```
Comment (muhammad-usama-sardar) [inline draft-fossati-tls-attestation.md:254]: Why not simply "Mutual attestation"?

File: diagrams/passport-tls-handshake.svg
Comment (muhammad-usama-sardar) [inline diagrams/passport-tls-handshake.svg:1]: What exactly does Server represent? Is it trusted/untrusted?

File: draft-fossati-tls-attestation.md (hunk 15 above)
Comment (yaronf) [reply draft-fossati-tls-attestation.md:254]: @muhammad-usama-sardar Because from a TLS point of view this is authentication.

File: draft-fossati-tls-attestation.md (hunk 5 above)
Comment (thomas-fossati) [inline draft-fossati-tls-attestation.md:241]:
```suggestion
When using the passport model, the remote attestation results obtained by the
attester from its trusted verifiers can be cached and used for any number of
subsequent TLS handshakes, as long as the freshness policy requirements are satisfied. ```

File: diagrams/passport-tls-handshake.svg
Comment (ionut-arm) [reply diagrams/passport-tls-handshake.svg:1]: I've made those changes to keep this diagram in line with the background check one in terminology. So the server and client are TLS server and TLS client. Realised pushing this up that there were still some gaps - in my local draft I've amended to have `Server | Attestation Service` in the same entity now.

File: draft-fossati-tls-attestation.md (hunk 5 above)
Comment (ionut-arm) [reply draft-fossati-tls-attestation.md:241]: Do I need to explain what the RP has to do to signal that these freshness requirements were not met?

File: draft-fossati-tls-attestation.md (hunk 14 above)
Comment (ionut-arm) [reply draft-fossati-tls-attestation.md:212]: I believe nonces are mentioned separately because they have to be passed along explicitly using the new extensions (for background check, at least), separate from the credentials.

File: draft-fossati-tls-attestation.md (hunk 5 above)
Comment (thomas-fossati) [reply draft-fossati-tls-attestation.md]: From their point of view, I don't think there's anything special: it's a failure in the credential verification.

**Comment by hannestschofenig** [comment]
This is a big PR. I would propose to merge it and then re-read the whole document to catch inconsistencies and bugs.

File: draft-fossati-tls-attestation.md (hunk 14 above)
Comment (muhammad-usama-sardar) [reply draft-fossati-tls-attestation.md:212]: Then probably define the "remote attestation credentials" such that it is clear that nonce is not a part of that.

**Comment by muhammad-usama-sardar** [comment]
> This is a big PR. I would propose to merge it and then re-read the whole document to catch inconsistencies and bugs.

Agree with Hannes. 
@ionut-arm can you share a PDF with all your local changes, via slack? I can offer to review the whole thing once.

**Comment by thomas-fossati** [comment]
> > This is a big PR. I would propose to merge it and then re-read the whole document to catch inconsistencies and bugs.
> 
> Agree with Hannes.

+1

**Comment by yogeshbdeshpande** [comment]
I agree with Hannes: We can refine further later but lets get this in, first with an attempt to incorporate the comments whatever we so far!

**Comment by ionut-arm** [comment]
Ok, let me merge, and feel free to review in `main`. 

> @ionut-arm can you share a PDF with all your local changes, via slack? I can offer to review the whole thing once.

I have no other local changes pending, everything's been pushed and rebased on top of the early exporter PR. But we can continue conversations on specific points in this PR if you'd like.

File: draft-fossati-tls-attestation.md (hunk 14 above)
Comment (ionut-arm) [reply draft-fossati-tls-attestation.md:213]: Do you mean in the `Conventions and Terminology` section or in this paragraph? I don't think nonces are usually lumped into "credentials", though perhaps this is confusing since the nonces are incorporated in credentials.
